File: Depth Camera/java/src/main/java/frc/robot/subsystems/DepthCamera.java
package frc.robot.subsystems;

import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
// Imports
import edu.wpi.first.wpilibj2.command.SubsystemBase;
import com.orbbec.obsensor.*;

public class DepthCamera extends SubsystemBase
{
    /**
     * Object Declarations
     */
    private OBContext obContext;
    private final Object pipeLock = new Object();
    private Pipeline pipeline;

    /**
     * Subsystem Constructor
     */
    public DepthCamera()
    {
        initCamera();
    }

    /**
     * Setup the Orbbec SDK and register the device connection status callback
     */
    private void initCamera()
    {
        // Init ObContext and regist the device status listnener
        obContext = new OBContext(new DeviceChangedCallback(){
            @Override
            public void onDeviceAttach(DeviceList deviceList)
            {
                // New devices attached
                // init the pipeline
                initPipeline(deviceList);
                startStreams();

                // Release the device list.
                deviceList.close();
            }

            @Override
            public void onDeviceDetach(DeviceList deviceList)
            {
                // Devices detached
                // Release the pipeline when the device is detached
                stopStreams();
                deInitPipeline();

                // Release the device list.
                deviceList.close();
            }
        });

        DeviceList deviceList = obContext.queryDevices();
        if (null != deviceList)
        {
            if (deviceList.getDeviceCount() > 0)
            {
                initPipeline(deviceList);
                startStreams();
            }
            deviceList.close();
        }
    }

    /**
     * Release the cameara when not used anymore
     */
    private void destroyCamera()
    {
        try 
        {
            if (null != obContext)
            {
                obContext.close();
                obContext = null;
            }
        } 
        catch(Exception e)
        {
            e.printStackTrace();
        }
    }

    /**
     * Init the pipeline
     * 
     * @param deviceList new attached device list
     */
    private void initPipeline(DeviceList deviceList)
    {
        synchronized(pipeLock) 
        {
            if (null != pipeline)
            {
                pipeline.close();
            }

            // Create the new attached device, default for index 0
            try(Device device = deviceList.getDevice(0))
            {
                pipeline = new Pipeline(device);
            }
            catch(Exception e)
            {
                e.printStackTrace();
            }
        }
    }


    /**
     * Release the pipeline
     */
    private void deInitPipeline()
    {
        synchronized(pipeLock)
        {
            try
            {
                if (null != pipeline)
                {
                    pipeline.close();
                    pipeline = null;
                }
            }
            catch (Exception e)
            {
                e.printStackTrace();
            }
        }
    }

    /**
     * Start streams, this function shows for streams config and starting streams by pipeline
     */
    public void startStreams()
    {
        synchronized(pipeLock)
        {
            if (null == pipeline)
            {
                System.out.println("Camera Failed! No device connected!");
                return;
            }

            // Create the config for starting streams
            Config config = new Config();

            // Config for color stream
            try (StreamProfileList colorProfileList = pipeline.getStreamProfileList(SensorType.COLOR))
            {
                // For example: You can also get StreamProfile this way.
                // for (int i = 0, N = colorProfileList.getStreamProfileCount(); i < N; i++) {
                //     VideoStreamProfile vsp = colorProfileList.getStreamProfile(i).as(StreamType.VIDEO);
                //     printVideoStreamProfile(vsp);
                // }

                // Filter out the StreamProfile you want based on specified arguments.
                //VideoStreamProfile colorProfile =  colorProfileList.getVideoStreamProfile(640, 480, Format.MJPG, 30);
                VideoStreamProfile colorProfile =  colorProfileList.getStreamProfile(0).as(StreamType.VIDEO);
                if (null == colorProfile) 
                {
                    throw new OBException("get color stream profile failed, no matched color stream profile!");
                }
                // print the color stream profile information.
                printVideoStreamProfile(colorProfile, SensorType.COLOR);
                // enable color stream
                config.enableStream(colorProfile);
                // release the color stream profile.
                colorProfile.close();
            }
            catch (Exception e)
            {
                e.printStackTrace();
            }

            // Config for depth stream.
            try (StreamProfileList depthProfileList = pipeline.getStreamProfileList(SensorType.DEPTH)) 
            {
                // For example: You can also get StreamProfile this way.
                // for (int i = 0, N = depthProfileList.getStreamProfileCount(); i < N; i++) {
                //     VideoStreamProfile vsp = depthProfileList.getStreamProfile(i).as(StreamType.VIDEO);
                //     printVideoStreamProfile(vsp);
                // }

                // Filter out the StreamProfile you want based on specified arguments.
                //VideoStreamProfile depthProfile = depthProfileList.getVideoStreamProfile(640, 480, Format.Y16, 30);
                VideoStreamProfile depthProfile = depthProfileList.getStreamProfile(0).as(StreamType.VIDEO);
                if (null == depthProfile) 
                {
                    throw new OBException("get depth stream profile failed, no matched depth stream profile!");
                }
                // print the depth stream profile information.
                printVideoStreamProfile(depthProfile, SensorType.DEPTH);
                // enable depth stream.
                config.enableStream(depthProfile);
                // release the depth stream profile.
                depthProfile.close();
            } 
            catch(Exception e) 
            {
                e.printStackTrace();
            }

            // Config for ir stream.
            try (StreamProfileList irProfileList = pipeline.getStreamProfileList(SensorType.IR)) 
            {
                // For example: You can also get StreamProfile this way.
                // for (int i = 0, N = irProfileList.getStreamProfileCount(); i < N; i++) {
                //     VideoStreamProfile vsp = irProfileList.getStreamProfile(i).as(StreamType.VIDEO);
                //     printVideoStreamProfile(vsp);
                // }

                // Filter out the StreamProfile you want based on specified arguments.
                // VideoStreamProfile irProfile = irProfileList.getVideoStreamProfile(640, 480, Format.Y16, 30);
                VideoStreamProfile irProfile = irProfileList.getStreamProfile(0).as(StreamType.VIDEO);
                if (null == irProfile) 
                {
                    throw new OBException("get ir stream profile failed, no matched ir streamm profile!");
                }
                // print the ir stream profile information.
                printVideoStreamProfile(irProfile, SensorType.IR);
                // enable ir stream.
                config.enableStream(irProfile);
                // release the ir stream profile.
                irProfile.close();
            } 
            catch(Exception e) 
            {
                e.printStackTrace();
            }

            // For example: You can enable hardware depth align to color if you want.
            // config.setAlignMode(AlignMode.ALIGN_D2C_HW_ENABLE);
            
            // For example: You can enable frame synchronization if you want.
            // pipeline.enableFrameSync();

            try
            {
                pipeline.start(config, new FrameSetCallback()
                {
                    @Override
                    public void onFrameSet(FrameSet frameSet) 
                    {
                        // process the frame set
                        processFrameSet(frameSet);
                        frameSet.close();
                    }
                });
                config.close();
            }
            catch (Exception e)
            {
                e.printStackTrace();
            }
        }
    }


    /**
     * Stop all the steams.
     */
    public void stopStreams()
    {
        synchronized(pipeLock)
        {
            try
            {
                if (null != pipeline)
                {
                    pipeline.stop();
                }
            }
            catch (Exception e)
            {
                e.printStackTrace();
            }
        }
    }

    /**
     * Print the video stream profile's information
     * 
     * @param vsp video stream profile
     */
    private void printVideoStreamProfile(VideoStreamProfile vsp, SensorType type)
    {
        StringBuilder sb = new StringBuilder()
        .append(" StreamType: " + vsp.getType() + "\n")
        .append(" Width: " + vsp.getWidth() + "\n")
        .append(" Height: " + vsp.getHeight() + "\n")
        .append(" FrameRate: " + vsp.getFrameRate() + "\n")
        .append(" Format: " + vsp.getFormat() + "\n");
        if (type == SensorType.COLOR)
            SmartDashboard.putString("Video Stream Profile Color", sb.toString());
        if (type == SensorType.DEPTH)
            SmartDashboard.putString("Video Stream Profile Depth", sb.toString());
        if (type == SensorType.IR)
            SmartDashboard.putString("Video Stream Profile IR", sb.toString());
    }

    /**
     * Print the video frame's information
     * 
     * @param vf video frame
     */
    private void printVideoFrame(VideoFrame vf, SensorType type)
    {
        StringBuilder sb = new StringBuilder()
        .append("FrameType: " + vf.getStreamType() + "\n")
        .append(" Width: " + vf.getWidth() + "\n")
        .append(" Height: " + vf.getHeight() + "\n")
        .append(" Format: " + vf.getFormat() + "\n")
        .append(" Index: " + vf.getFrameIndex() + "\n")
        .append(" Timestamp: " + vf.getTimeStamp() + "\n")
        .append(" SysTimestamp: " + vf.getSystemTimeStamp() + "\n");
        if (type == SensorType.COLOR)
            SmartDashboard.putString("Video Frame Color", sb.toString());
        if (type == SensorType.DEPTH)
            SmartDashboard.putString("Video Frame Depth", sb.toString());
        if (type == SensorType.IR)
            SmartDashboard.putString("Video Frame IR", sb.toString());
    }

    /**
     * Process the frame set
     * 
     * @param frameSet new frame set
     */
    private void processFrameSet(FrameSet frameSet)
    {
        // Get the color frame in frameSet.
        try (ColorFrame colorFrame = frameSet.getFrame(FrameType.COLOR)) 
        {
            if (null != colorFrame) 
            {
                // print the color frame's information.
                printVideoFrame(colorFrame, SensorType.COLOR);
            }
        } 
        catch(Exception e) 
        {
            e.printStackTrace();
        }
        // Get the depth frame in frameSet.
        try (DepthFrame depthFrame = frameSet.getFrame(FrameType.DEPTH)) 
        {
            if (null != depthFrame) 
            {
                // print the depth frame's information.
                printVideoFrame(depthFrame, SensorType.DEPTH);
            }
        } 
        catch(Exception e) 
        {
            e.printStackTrace();
        }
        // Get the ir frame in frameSet.
        try (IRFrame irFrame = frameSet.getFrame(FrameType.IR)) 
        {
            if (null != irFrame) 
            {
                // print the ir frame's information.
                printVideoFrame(irFrame, SensorType.IR);
            }
        } 
        catch(Exception e) 
        {
            e.printStackTrace();
        }
    }
}

File: Depth Camera/cpp/src/main/cpp/subsystems/DepthCamera.cpp
#include "subsystems/DepthCamera.h"
#include <iomanip>
#include <iostream>
#include <sstream>
#include <bitset>

std::ostream &operator<<(std::ostream &os, const OBFrameType frameType) {
    if(frameType == OB_FRAME_COLOR) {
        os << "OB_FRAME_COLOR";
    }
    else if(frameType == OB_FRAME_IR) {
        os << "OB_FRAME_IR";
    }
    else if(frameType == OB_FRAME_IR_LEFT) {
        os << "OB_FRAME_IR_LEFT";
    }
    else if(frameType == OB_FRAME_IR_RIGHT) {
        os << "OB_FRAME_IR_RIGHT";
    }
    else if(frameType == OB_FRAME_DEPTH) {
        os << "OB_FRAME_DEPTH";
    }
    else if(frameType == OB_FRAME_ACCEL) {
        os << "OB_FRAME_ACCEL";
    }
    else if(frameType == OB_FRAME_GYRO) {
        os << "OB_FRAME_GYRO";
    }
    else {
        os << (int)frameType;
    }
    return os;
}

std::ostream &operator<<(std::ostream &os, std::shared_ptr<ob::DeviceInfo> deviceInfo) {
    std::ostringstream oss;
    oss << std::setfill('0') << std::setw(4) << std::hex << "vid: 0x" << deviceInfo->vid() << ", pid: 0x" << deviceInfo->pid();
    os << oss.str() << ", sn: " << std::string(deviceInfo->serialNumber()) << ", uid: " << std::string(deviceInfo->uid());
    return os;
}

std::ostream &operator<<(std::ostream &os, std::shared_ptr<ob::DepthFrame> frame) {
    uint16_t *data       = (uint16_t *)frame->data();
    auto      width      = frame->width();
    auto      height     = frame->height();
    auto      scale      = frame->getValueScale();
    uint16_t  pixelValue = *(data + width * height / 2 + width / 2);
    os << "FrameType: " << frame->type() << ", index: " << frame->index() << ", width: " << width << ", height: " << height << ", format: " << frame->format()
       << ", timeStampUs: " << frame->timeStampUs() << "us, centerDepth: " << (float)pixelValue * scale << "mm";
    return os;
}

std::ostream &operator<<(std::ostream &os, std::shared_ptr<ob::VideoFrame> frame) {
    os << "FrameType: " << frame->type() << ", index: " << frame->index() << ", width: " << frame->width() << ", height: " << frame->height()
       << ", format: " << frame->format() << ", timeStampUs: " << frame->timeStampUs() << "us";
    return os;
}

std::ostream &operator<<(std::ostream &os, std::shared_ptr<ob::VideoStreamProfile> vsp) {
    os << "StreamType:" << vsp->type() << ", Width: " << vsp->width() << ", Height: " << vsp->height() << ", FrameRate: " << vsp->fps()
       << ", format:" << vsp->format();
    return os;
}

DepthCamera::DepthCamera() {
    setupOrbbecSDK();
}

DepthCamera::~DepthCamera() {
    teardownOrbbecSDK();
}

void DepthCamera::Periodic() {
    // std::cout << "Printing here would flood the console output window" << std::endl;
}

void DepthCamera::setupOrbbecSDK() {
    // Print the sdk version number, the sdk version number is divided into major version number, minor version number and revision number
    std::cout << "SDK version: " << ob::Version::getMajor() << "." << ob::Version::getMinor() << "." << ob::Version::getPatch() << std::endl;
    // Print sdk stage version
    std::cout << "SDK stage version: " << ob::Version::getStageVersion() << std::endl;
    obContex_ = std::make_shared<ob::Context>();
    obContex_->setDeviceChangedCallback([&](std::shared_ptr<ob::DeviceList> removedList, std::shared_ptr<ob::DeviceList> addedList) {
        handleDeviceDisconnected(removedList);
        handleDeviceConnected(addedList);
    });
    handleDeviceConnected(obContex_->queryDeviceList());
}

void DepthCamera::teardownOrbbecSDK() {
    for(const auto &pair: pipelineHolderMap_) {
        stopStream(pair.second);
        std::cout << "teardownOrbbecSDK: stop streams for device: " << pair.second->deviceInfo << std::endl;
    }
    pipelineHolderMap_.clear();
    obContex_.reset();
}

// Device connection callback
void DepthCamera::handleDeviceConnected(std::shared_ptr<ob::DeviceList> connectList) {
    std::lock_guard<std::recursive_mutex> lk(pipelineHolderMutex_);
    const auto                            deviceCount = connectList->deviceCount();
    std::cout << "Device connect, deviceCount: " << deviceCount << std::endl;
    for(uint32_t i = 0; i < deviceCount; i++) {
        auto uid = std::string(connectList->uid(i));
        auto itr = pipelineHolderMap_.find(uid);
        if(itr != pipelineHolderMap_.end()) {
            std::cout << "Device connected. device already exit. " << itr->second->deviceInfo << std::endl;
        }
        else {
            auto device     = connectList->getDevice(i);
            auto deviceInfo = device->getDeviceInfo();

            std::shared_ptr<ob::Pipeline>   pipeline(new ob::Pipeline(device));
            std::shared_ptr<PipelineHolder> holder(new PipelineHolder{ pipeline, deviceInfo, false });
            pipelineHolderMap_.insert({ uid, holder });

            std::cout << "Device connected. " << deviceInfo << std::endl;
            startStream(holder);
        }
    }
}

// Device disconnect callback
void DepthCamera::handleDeviceDisconnected(std::shared_ptr<ob::DeviceList> disconnectList) {
    std::lock_guard<std::recursive_mutex> lk(pipelineHolderMutex_);
    const auto                            deviceCount = disconnectList->deviceCount();
    std::cout << "Device disconnect, deviceCount: " << deviceCount << std::endl;
    for(uint32_t i = 0; i < deviceCount; i++) {
        auto uid = std::string(disconnectList->uid(i));
        auto itr = pipelineHolderMap_.find(uid);
        if(itr != pipelineHolderMap_.end()) {
            auto deviceInfo = itr->second->deviceInfo;
            stopStream(itr->second);
            pipelineHolderMap_.erase(uid);
            std::cout << "Device disconnected. " << deviceInfo << std::endl;
        }
        else {
            std::cout << "Device disconnect, unresolve deviceUid: " << uid << std::endl;
        }
    }
}

void DepthCamera::rebootDevices() {
    std::cout << "Reboot devices, count: " << pipelineHolderMap_.size() << std::endl;
    for(auto itr = pipelineHolderMap_.begin(); itr != pipelineHolderMap_.end(); itr++) {
        try {
            std::cout << "Reboot device, " << itr->second->deviceInfo << std::endl;
            stopStream(itr->second);
            itr->second->pipeline->getDevice()->reboot();
        }
        catch(ob::Error &e) {
            std::cerr << "Reboot device failed. function:" << e.getName() << "\nargs:" << e.getArgs() << "\nmessage:" << e.getMessage()
                      << "\ntype:" << e.getExceptionType() << std::endl;
        }
    }
}

void DepthCamera::processFrameset(std::string deviceSN, std::shared_ptr<FramePrintInfo> printInfo, std::shared_ptr<ob::FrameSet> frameSet) {
    // Get the depth data frame
    auto depthFrame = frameSet->depthFrame();
    if(depthFrame) {
        printInfo->depthCount++;
        if(printInfo->depthCount == 15) {
            std::cout << "=====Depth Frame Info====== SN: " << std::string(deviceSN) << ", " << depthFrame << std::endl;
            printInfo->depthCount = 0;
        }
    }

    // Get the ir data frame
    auto irFrame = frameSet->irFrame();
    if(irFrame) {
        printInfo->irCount++;
        if(printInfo->irCount == 15) {
            std::cout << "=====IR Frame Info====== SN: " << std::string(deviceSN) << ", " << std::dynamic_pointer_cast<ob::VideoFrame>(irFrame) << std::endl;
            printInfo->irCount = 0;
        }
    }

    // Get the ir left data frame
    auto irLeftFrame = frameSet->getFrame(OB_FRAME_IR_LEFT);
    if(irLeftFrame) {
        printInfo->irLeftCount++;
        if(printInfo->irLeftCount == 15) {
            std::cout << "=====IR Left Frame Info====== SN: " << std::string(deviceSN) << ", " << std::dynamic_pointer_cast<ob::VideoFrame>(irLeftFrame)
                      << std::endl;
            printInfo->irLeftCount = 0;
        }
    }

    // Get the ir right data frame
    auto irRightFrame = frameSet->getFrame(OB_FRAME_IR_RIGHT);
    if(irRightFrame) {
        printInfo->irRightCount++;
        if(printInfo->irRightCount == 15) {
            std::cout << "=====IR Right Frame Info====== SN: " << std::string(deviceSN) << ", " << std::dynamic_pointer_cast<ob::VideoFrame>(irRightFrame)
                      << std::endl;
            printInfo->irRightCount = 0;
        }
    }

    // Get the color data frame
    auto colorFrame = frameSet->colorFrame();
    if(colorFrame) {
        printInfo->colorCount++;
        if(printInfo->colorCount == 15) {
            std::cout << "=====Color Frame Info====== SN: " << std::string(deviceSN) << ", " << std::dynamic_pointer_cast<ob::VideoFrame>(colorFrame)
                      << std::endl;
            printInfo->colorCount = 0;
        }
    }
}

void DepthCamera::startStream(std::shared_ptr<PipelineHolder> holder) {
    std::shared_ptr<FramePrintInfo> printInfo(new FramePrintInfo{});
    std::string                     deviceSN = std::string(holder->deviceInfo->serialNumber());
    ob::FrameSetCallback callback = [&, deviceSN, printInfo](std::shared_ptr<ob::FrameSet> frameSet) { processFrameset(deviceSN, printInfo, frameSet); };

    // Start video stream according to the stream profile of the configuration file.If there is no configuration file, the first stream profile will be used.
    try {
        std::cout << "startStream " << holder->deviceInfo << std::endl;
        // Create the config for starting streams.
        std::shared_ptr<ob::Config> config = std::make_shared<ob::Config>();
        // Config for color stream.
        try {
            auto colorProfileList = holder->pipeline->getStreamProfileList(OB_SENSOR_COLOR);
            // For example: You can also get StreamProfile this way.
            // for (int i = 0, N = colorProfileList->count(); i < N; i++) {
            //     auto vsp = colorProfileList->getProfile(i)->as<ob::VideoStreamProfile>();
            //     std::cout << vsp << std::endl;
            // }

            // Filter out the StreamProfile you want based on specified arguments.
            // auto colorProfile =  colorProfileList->getVideoStreamProfile(640, 480, OB_FORMAT_MJPEG, 30);
            auto colorProfile = colorProfileList->getProfile(0)->as<ob::VideoStreamProfile>();
            std::cout << "ColorProfile: " << colorProfile << std::endl;
            config->enableStream(colorProfile);
        }
        catch(...) {
            std::cerr << "Config for color profile failed!" << std::endl;
        }

        // Config for depth stream.
        try {
            auto depthProfileList = holder->pipeline->getStreamProfileList(OB_SENSOR_DEPTH);
            // For example: You can also get StreamProfile this way.
            // for (int i = 0, N = depthProfileList->count(); i < N; i++) {
            //     auto vsp = depthProfileList->getProfile(i)->as<ob::VideoStreamProfile>();
            //     std::cout << vsp << std::endl;
            // }

            // Filter out the StreamProfile you want based on specified arguments.
            // auto depthProfile =  depthProfileList->getVideoStreamProfile(640, 480, Format.Y16, 30);
            auto depthProfile = depthProfileList->getProfile(0)->as<ob::VideoStreamProfile>();
            std::cout << "DepthProfile: " << depthProfile << std::endl;
            config->enableStream(depthProfile);
        }
        catch(...) {
            std::cerr << "Config for depth profile failed!" << std::endl;
        }

        // Config for ir stream.
        try {
            auto irProfileList = holder->pipeline->getStreamProfileList(OB_SENSOR_IR);
            // For example: You can also get StreamProfile this way.
            // for (int i = 0, N = irProfileList->count(); i < N; i++) {
            //     auto vsp = irProfileList->getProfile(i)->as<ob::VideoStreamProfile>();
            //     std::cout << vsp << std::endl;
            // }

            // Filter out the StreamProfile you want based on specified arguments.
            // auto irProfile =  depthProfileList->getVideoStreamProfile(640, 480, Format.Y16, 30);
            auto irProfile = irProfileList->getProfile(0)->as<ob::VideoStreamProfile>();
            std::cout << "IrProfile: " << irProfile << std::endl;
            config->enableStream(irProfile);
        }
        catch(...) {
            std::cerr << "Config for ir profile failed!" << std::endl;
        }

        holder->pipeline->start(config, callback);
        holder->isStarted = true;
    }
    catch(...) {
        std::cout << "Pipeline start failed!" << std::endl;
        holder->isStarted = false;
    }
}

void DepthCamera::stopStream(std::shared_ptr<PipelineHolder> holder) {
    if(!holder->isStarted) {
        return;
    }

    try {
        std::cout << "stopStream " << holder->deviceInfo << std::endl;
        holder->isStarted = false;
        holder->pipeline->stop();
    }
    catch(ob::Error &e) {
        std::cerr << "stopStream failed., function:" << e.getName() << "\nargs:" << e.getArgs() << "\nmessage:" << e.getMessage()
                  << "\ntype:" << e.getExceptionType() << std::endl;
    }
}
